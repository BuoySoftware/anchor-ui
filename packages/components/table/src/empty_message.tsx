import { useTranslation } from "react-i18next";
import { Body } from "@buoysoftware/anchor-typography";
import { useTable, DEFAULT_TABLE_NAMESPACE } from "./use_table";

export const EmptyMessage: React.FC = (): React.ReactElement => {
  const { tNamespace, tableName } = useTable();
  const { t } = useTranslation();

  const translationLookup = [
    `${tNamespace}:${tableName}.empty_message`,
    `${tNamespace}:tables.${tableName}.empty_message`,
    `${DEFAULT_TABLE_NAMESPACE}:${tableName}.empty_message`,
    `${DEFAULT_TABLE_NAMESPACE}:empty_message`,
  ];

  return (
    <Body mt="xl" size="m" textAlign="center" data-testid="empty-table-state">
      {t(translationLookup)}
    </Body>
  );
};
