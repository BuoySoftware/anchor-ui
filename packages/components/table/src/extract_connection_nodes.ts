export interface Edge<NodeData> {
  node: NodeData;
}
export interface Connection<NodeData> {
  edges: Edge<NodeData>[];
}

export function extractConnectionNodes<NodeData>(
  connection?: Connection<NodeData>
): NodeData[] {
  if (!connection) return [];

  return connection.edges.map(({ node }) => node);
}
