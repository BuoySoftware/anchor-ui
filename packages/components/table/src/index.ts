export * from "./empty_cell";
export * from "./table";
