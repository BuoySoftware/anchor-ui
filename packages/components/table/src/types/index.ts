import type { PropertyPath } from "lodash";
import type { UniqueIdentifier } from "@dnd-kit/core";
import { BodyProps } from "@buoysoftware/anchor-typography";

// Pagination utility types
export interface Edge<NodeData> {
  node: Record<NodeData>;
}

export interface Connection<NodeData> {
  edges: Edge<NodeData>[];
}

export interface ConnectionWithEdges<EdgeData> {
  edges: EdgeData[];
}

export interface PageInfo {
  startCursor: string | null;
  endCursor: string | null;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  perPage: number;
}

export interface PageableConnection {
  pageInfo: PageInfo;
  totalCount: number;
}

export interface PaginationVariables {
  variables: {
    after?: string;
    before?: string;
    first?: number;
    last?: number;
  };
}

export type FetchMore = (
  fetchMoreOptions: PaginationVariables
) => Promise<unknown>;

export type PaginationData<RecordData> = (
  | Connection<RecordData>
  | ConnectionWithEdges<RecordData>
) &
  PageableConnection;

// Table component Types
export type CellProps = Partial<BodyProps> | undefined;

export type Record<RecordData> = RecordData & { id?: string };

export type RowAction<RecordData> = (rowData: RecordData) => void;

export type TableVariant = "default" | "condensed";

export interface TableCellConfig<RecordData> {
  cellProps?: (rowData: RecordData) => CellProps;
  dataKey: PropertyPath;
  name?: string;
  render?: (rowData: RecordData) => React.ReactNode;
}

export interface TableProps<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  TableRowComponent?: React.FC<TableRow<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  fetchMore?: FetchMore;
  name: string;
  paginationData?: PaginationData<RecordData>;
  recordIdKey: PropertyPath;
  records: Record<RecordData>[];
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
  tNamespace?: string;
}

export interface TableBodyProps<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  TableRowComponent?: React.FC<TableRow<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  recordIdKey: PropertyPath;
  records: Record<RecordData>[];
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
}

export type RenderPlaceholder<RecordData> = (
  rowData: RecordData
) => React.ReactElement | string;

export interface TableRow<RecordData> {
  TableCellComponent?: React.FC<TableCell<RecordData>>;
  cellConfigs: TableCellConfig<RecordData>[];
  record: Record<RecordData>;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  rowAction?: RowAction<RecordData>;
  rowId: string;
}

export interface TableCell<RecordData> {
  cellConfig: TableCellConfig<RecordData>;
  record: RecordData;
  renderPlaceholder?: RenderPlaceholder<RecordData>;
  testId: string;
  variant?: TableVariant;
}

// Sortable table types
export interface SortableRecord extends Object {
  id: UniqueIdentifier;
}

export interface SortableTableCellConfig<RecordData>
  extends TableCellConfig<RecordData> {
  sortHandle?: boolean;
}

export interface SortableTableRow<RecordData>
  extends Omit<TableRow<RecordData>, "cellConfigs"> {
  cellConfigs: SortableTableCellConfig<RecordData>[];
  sortHandle?: boolean;
}

export interface SortableTableCell<RecordData>
  extends Omit<TableCell<RecordData>, "cellConfig"> {
  cellConfig: SortableTableCellConfig<RecordData>;
  sortHandle?: boolean;
}

export interface SortableTableProps<RecordData>
  extends Omit<
    TableProps<RecordData>,
    "cellConfigs" | "TableCellComponent" | "TableRowComponent"
  > {
  cellConfigs: SortableTableCellConfig<RecordData>[];
  TableCellComponent?: React.FC<SortableTableCell<RecordData>>;
  TableRowComponent?: React.FC<SortableTableRow<RecordData>>;
  onSort?: (records: Record<RecordData>[]) => void;
}
