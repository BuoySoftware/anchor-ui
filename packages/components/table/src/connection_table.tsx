import { Table, TableProps } from "./table";
import { Connection, PageableConnection } from "./types";
import { extractConnectionNodes } from "./extract_connection_nodes";

interface OwnProps<RecordData> {
  connection?: Connection<RecordData> & PageableConnection;
}

type ConnectionTableProps<RecordData> = OwnProps<RecordData> &
  Omit<TableProps<RecordData>, "records" | "paginationData">;

export const ConnectionTable = <RecordData,>({
  connection,
  ...tableBuilderProps
}: ConnectionTableProps<RecordData>): React.ReactElement => {
  const records = extractConnectionNodes(connection);

  return (
    <Table
      paginationData={connection}
      records={records}
      {...tableBuilderProps}
    />
  );
};
