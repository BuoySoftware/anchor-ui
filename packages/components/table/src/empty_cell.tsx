import { TableCellConfig } from "./types";

export const EMPTY_CELL: TableCellConfig<any> = {
  dataKey: "",
  render: () => <></>,
};
