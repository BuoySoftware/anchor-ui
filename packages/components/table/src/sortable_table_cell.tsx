import { useSortable } from "@dnd-kit/sortable";
import { useTheme } from "@buoysoftware/anchor-theme"

import sortHandleGraphic from "./sort_handle.svg";
import { SortableRecord, SortableTableCell as GeneralSortableTableCell } from "./types";
import { Td } from "./td";

export type SortableTableCellProps<RecordData> = GeneralSortableTableCell<RecordData>;

export const SortableTableCell = function <RecordData extends SortableRecord>({
  record,
  cellConfig: sortableCellConfig,
  ...tableCellProps
}: SortableTableCellProps<RecordData>): React.ReactElement {
  const { attributes, listeners, setActivatorNodeRef } = useSortable({
    id: record.id
  });
  const { sizes, space } = useTheme();
  const { sortHandle, ...cellConfig } = sortableCellConfig;

  const sortHandleProps = {
    background: `${space.xs} center url(${sortHandleGraphic}) no-repeat`,
    minHeight: "sortHandle",
    minWidth: "sortHandle",
    pl: `calc(${space.xs} + ${sizes.sortHandle} + ${space.xs})`,
    ref: setActivatorNodeRef,
    style: {
      cursor: "grab"
    },
    ...attributes,
    ...listeners
  }

  return (
    <Td
      cellConfig={cellConfig}
      record={record}
      {...(sortHandle ? sortHandleProps : undefined)}
      {...tableCellProps}
    />
  );
};
