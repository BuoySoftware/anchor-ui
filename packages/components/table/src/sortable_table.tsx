import { useEffect, useCallback, useState } from "react";
import {
  DndContext,
  DragEndEvent,
  KeyboardSensor,
  PointerSensor,
  UniqueIdentifier,
  closestCenter,
  useSensor,
  useSensors,
} from "@dnd-kit/core";
import {
  restrictToVerticalAxis,
  restrictToWindowEdges,
} from "@dnd-kit/modifiers";
import {
  SortableContext,
  arrayMove,
  sortableKeyboardCoordinates,
  verticalListSortingStrategy,
} from "@dnd-kit/sortable";

import {
  SortableRecord,
  SortableTableProps as GeneralSortableTable,
} from "./types";
import { SortableTableCell } from "./sortable_table_cell";
import { SortableTableRow } from "./sortable_table_row";
import { Table } from "./table";

type SortableTableProps<RecordData> = GeneralSortableTable<RecordData>;

export const SortableTable = <RecordData extends SortableRecord>({
  onSort,
  records,
  ...tableBuilderProps
}: SortableTableProps<RecordData>): React.ReactElement => {
  const [sortedRecords, setSortedRecords] = useState(records);
  const sensors = useSensors(
    useSensor(PointerSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates })
  );

  const handleDragEnd = useCallback(
    ({ active, over }: DragEndEvent) => {
      if (!active.id || !over?.id) return;
      if (active.id === over.id) return;

      function indexById(records: RecordData[], id: UniqueIdentifier) {
        const record = records.find((record) => id === record.id);
        return record ? records.indexOf(record) : -1;
      }

      setSortedRecords((sortedRecords) => {
        const oldIndex = indexById(sortedRecords, active.id);
        const newIndex = indexById(sortedRecords, over.id);
        const resortedRecords = arrayMove(sortedRecords, oldIndex, newIndex);

        return resortedRecords;
      });
    },
    [setSortedRecords]
  );

  useEffect(() => {
    if (!onSort) return;

    if (sortedRecords !== records) {
      onSort(sortedRecords);
    }
  }, [onSort, sortedRecords, records]);

  return (
    <DndContext
      collisionDetection={closestCenter}
      modifiers={[restrictToVerticalAxis, restrictToWindowEdges]}
      onDragEnd={handleDragEnd}
      sensors={sensors}
    >
      <SortableContext
        items={sortedRecords}
        strategy={verticalListSortingStrategy}
      >
        <Table
          TableCellComponent={SortableTableCell}
          TableRowComponent={SortableTableRow}
          records={sortedRecords}
          variant="sortable"
          {...tableBuilderProps}
        />
      </SortableContext>
    </DndContext>
  );
};
