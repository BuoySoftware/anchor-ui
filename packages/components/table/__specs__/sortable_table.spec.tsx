import i18n from "i18next";
import userEvent from "@testing-library/user-event";
import { SortableTable } from "../src/sortable_table";
import { SortableTableCellConfig } from "../src/types";
import { act, render } from "../test_utils";
import { initReactI18next } from "react-i18next";

interface Donation {
  __typename: string;
  id: string;
  status: string;
}

const cellConfigs: SortableTableCellConfig<Donation>[] = [
  { dataKey: "id", sortHandle: true },
  { dataKey: "status" },
];

const records: Donation[] = [
  {
    __typename: "Donation",
    id: "1",
    status: "Good",
  },
  {
    __typename: "Donation",
    id: "2",
    status: "Bad",
  },
  {
    __typename: "Donation",
    id: "3",
    status: "Ugly",
  },
  {
    __typename: "Donation",
    id: "4",
    status: "Other",
  },
];

describe("<SortableTable />", () => {
  const tables = {
    donations: {
      th: {
        id: "ID",
        status: "Status",
      },
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["tables"],
    resources: {
      en: {
        tables,
      },
    },
  });

  it("renders a sortable table", () => {
    const component = render(
      <SortableTable
        cellConfigs={cellConfigs}
        name="donations"
        onSort={jest.fn()}
        recordIdKey="id"
        records={records}
      />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  describe("sorting by mouse", () => {
    it("fires callback with sorted items", async () => {
      const user = userEvent.setup();
      const onSort = jest.fn();
      const component = render(
        <SortableTable
          cellConfigs={cellConfigs}
          name="donations"
          onSort={onSort}
          recordIdKey="id"
          records={records}
        />
      );
      const rows = component.queryAllByRole("button");

      await act(() => user.pointer({ keys: "[MouseLeft>]", target: rows[3] }))
      await act(() => user.pointer({ keys: "[/MouseLeft]", target: rows[0] }))

      expect(onSort).toHaveBeenCalledWith([
        records[3],
        records[0],
        records[1],
        records[2]
      ])
    })
  });

  describe("sorting by keyboard", () => {
    it("fires callback with sorted items", async () => {
      const user = userEvent.setup();
      const onSort = jest.fn();
      render(
        <SortableTable
          cellConfigs={cellConfigs}
          name="donations"
          onSort={onSort}
          recordIdKey="id"
          records={records}
        />
      );

      await act(() => user.keyboard("{Shift>}[Tab]{/Shift}")) // Focus last row
      await act(() => user.keyboard("[Space]")) // Activate "Drag" Start
      await act(() => user.keyboard("[ArrowUp]")) // "Drag" up
      await act(() => user.keyboard("[ArrowUp]")) // "Drag" up
      await act(() => user.keyboard("[ArrowUp]")) // "Drag" up
      await act(() => user.keyboard("[Space]")) // End "Drag"

      expect(onSort).toHaveBeenCalledWith([
        records[3],
        records[0],
        records[1],
        records[2]
      ])
    })
  });
});
