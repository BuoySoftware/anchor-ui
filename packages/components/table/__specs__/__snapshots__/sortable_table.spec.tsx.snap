// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SortableTable /> renders a sortable table 1`] = `
<DocumentFragment>
  .c2 {
  box-sizing: border-box;
  color: #101112;
  width: s;
  height: s;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
}

.c5 {
  box-sizing: border-box;
  background: 0.5rem center url(test-file-stub) no-repeat;
  color: #101112;
  width: m;
  height: m;
  min-height: 16px;
  min-width: 16px;
  padding-left: calc(0.5rem + 16px + 0.5rem);
  padding-right: 0.5rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
}

.c7 {
  box-sizing: border-box;
  color: #101112;
  width: m;
  height: m;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
}

.c3 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: 0.875rem;
  font-weight: 600;
  line-height: 1rem;
}

.c6 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-style: none;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1rem;
}

.c0 {
  width: 100%;
  box-sizing: border-box;
}

.c0 tr:has(td:focus-visible) {
  outline: 2px solid #276EF1;
  border-radius: 4px;
}

.c0 tr:has(td:focus-visible) td {
  outline: none;
}

.c4 tr {
  border-bottom: 1px solid #BABEC3;
}

.c1 {
  border-bottom: 1px solid #BABEC3;
}

.c1 th {
  overflow: hidden;
  text-align: left;
  text-overflow: ellipsis;
  vertical-align: bottom;
}

.c1 th:last-child {
  padding-right: 0;
}

.c1:first-child th {
  padding-top: 0;
}

<table
    class="c0"
    data-testid="donations-table"
  >
    <thead
      class="c1"
    >
      <tr>
        <th
          class="c2 c3"
          color="text.primary"
          font-size="font-size-100"
          font-weight="600"
        >
          ID
        </th>
        <th
          class="c2 c3"
          color="text.primary"
          font-size="font-size-100"
          font-weight="600"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody
      class="c4"
    >
      <tr
        class="sortable-table"
        data-testid="tr-1"
      >
        <td
          aria-describedby="DndDescribedBy-0"
          aria-disabled="false"
          aria-roledescription="sortable"
          class="c5 c6"
          color="text.primary"
          data-testid="tr-1-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
          role="button"
          style="cursor: grab;"
          tabindex="0"
        >
          1
        </td>
        <td
          class="c7 c6"
          color="text.primary"
          data-testid="tr-1-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Good
        </td>
      </tr>
      <tr
        class="sortable-table"
        data-testid="tr-2"
      >
        <td
          aria-describedby="DndDescribedBy-0"
          aria-disabled="false"
          aria-roledescription="sortable"
          class="c5 c6"
          color="text.primary"
          data-testid="tr-2-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
          role="button"
          style="cursor: grab;"
          tabindex="0"
        >
          2
        </td>
        <td
          class="c7 c6"
          color="text.primary"
          data-testid="tr-2-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Bad
        </td>
      </tr>
      <tr
        class="sortable-table"
        data-testid="tr-3"
      >
        <td
          aria-describedby="DndDescribedBy-0"
          aria-disabled="false"
          aria-roledescription="sortable"
          class="c5 c6"
          color="text.primary"
          data-testid="tr-3-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
          role="button"
          style="cursor: grab;"
          tabindex="0"
        >
          3
        </td>
        <td
          class="c7 c6"
          color="text.primary"
          data-testid="tr-3-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Ugly
        </td>
      </tr>
      <tr
        class="sortable-table"
        data-testid="tr-4"
      >
        <td
          aria-describedby="DndDescribedBy-0"
          aria-disabled="false"
          aria-roledescription="sortable"
          class="c5 c6"
          color="text.primary"
          data-testid="tr-4-td-id"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
          role="button"
          style="cursor: grab;"
          tabindex="0"
        >
          4
        </td>
        <td
          class="c7 c6"
          color="text.primary"
          data-testid="tr-4-td-status"
          font-size="font-size-100"
          font-style="none"
          font-weight="400"
        >
          Other
        </td>
      </tr>
    </tbody>
  </table>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</DocumentFragment>
`;
